File: programming 1/lab 5/src/models/Location.java
package models;

public class Location implements Comparable<Location> {
    private Double x; //Поле не может быть null
    private long y;
    private Float z; //Поле не может быть null
    private String name; //Поле не может быть null

    public void setName(String name) {
        if (name.trim().isEmpty()) throw new NullPointerException("Location.name: Значение не может быть пустым!");
        this.name = name;
    }

    public void setY(String line) {
        if (line.trim().isEmpty()) {
            this.y = 0;
            return;
        }
        try {
            this.y = Long.parseLong(line);
        }
        catch (NumberFormatException e) {
            throw new NumberFormatException("Location.y: Неверный формат ввода! Требуется числовое значение.");
        }
    }

    public void setX(String line) {
        if (line.trim().isEmpty()) throw new NullPointerException("Location.x: Значение не может быть пустым!");
        try {
            this.x = Double.parseDouble(line);
        }
        catch (NumberFormatException e) {
            throw new NumberFormatException("Location.x: Неверный формат ввода! Требуется числовое значение.");
        }
    }

    public void setZ(String line) {
        if (line.trim().isEmpty()) throw new NullPointerException("Location.z: Значение не может быть пустым!");
        try {
            this.z = Float.parseFloat(line);
        }
        catch (NumberFormatException e) {
            throw new NumberFormatException("Location.z: Неверный формат ввода! Требуется числовое значение.");
        }
    }

    public Double getX() {
        return x;
    }

    public Float getZ() {
        return z;
    }

    public long getY() {
        return y;
    }

    public String getName() {
        return name;
    }

    @Override
    public String toString() {
        return "Location{" +
                "name='" + name +
                ", x=" + x +
                ", y=" + y +
                ", z=" + z + '\'' +
                '}';
    }

    @Override
    public int compareTo(Location location) {
        if (Double.compare(x, location.x) != 0) {
            return Double.compare(x, location.x);
        }
        else if (y != location.y) {
            return Long.compare(y, location.y);
        }
        else if (Float.compare(z, location.z) != 0) {
            return Float.compare(z, location.z);
        }
        return name.compareTo(location.name);
    }
}

File: programming 1/lab 6/kotlin/lab 6/src/main/java/common/models/Location.kt
package common.models

import java.io.Serializable

class Location : Comparable<Location?>, Serializable {
    var x: Double? = null //Поле не может быть null
    var y: Long = 0
    var z: Float? = null //Поле не может быть null
    var name: String? = null //Поле не может быть null
        set(value) {
            if (value == null || value.trim().isEmpty()) throw NullPointerException("Location.name: Значение не может быть пустым!")
            else field = value
        }

    fun setY(line: String) {
        if (line.trim { it <= ' ' }.isEmpty()) {
            this.y = 0
            return
        }
        try {
            this.y = line.toLong()
        } catch (e: NumberFormatException) {
            throw NumberFormatException("Location.y: Неверный формат ввода! Требуется числовое значение.")
        }
    }

    fun setX(line: String) {
        if (line.trim().isEmpty()) throw NullPointerException("Location.x: Значение не может быть пустым!")
        try {
            this.x = line.toDouble()
        } catch (e: NumberFormatException) {
            throw NumberFormatException("Location.x: Неверный формат ввода! Требуется числовое значение.")
        }
    }

    fun setZ(line: String) {
        if (line.trim().isEmpty()) throw NullPointerException("Location.z: Значение не может быть пустым!")
        try {
            this.z = line.toFloat()
        } catch (e: NumberFormatException) {
            throw NumberFormatException("Location.z: Неверный формат ввода! Требуется числовое значение.")
        }
    }

    override fun toString(): String {
        return "Location{" +
                "name='" + name +
                ", x=" + x +
                ", y=" + y +
                ", z=" + z + '\'' +
                '}'
    }

    override fun compareTo(other: Location?): Int {
        if (x!!.compareTo(other!!.x!!) != 0) {
            return x!!.compareTo(other.x!!)
        } else if (y != other.y) {
            return y.compareTo(other.y)
        } else if (z!!.compareTo(other.z!!) != 0) {
            return z!!.compareTo(other.z!!)
        }
        return name!!.compareTo(other.name!!)
    }
}
